### PR TITLE
GH1156: Fixes assembly loading on .NET Core.

### LINKED ISSUE #1156: CoreCLR: Can't load addins
Need to implement a custom assembly loader for this.

### PR DESCRIPTION
**Pull request by patriksvensson**
Closes #1156

### CODE REVIEW

File: src/Cake/Program.cs [hunk 1]
```diff
@@ -86,6 +86,6 @@ public static int Main()
                 }
                 return 1;
             }
-        }
+         }
```
Comment (devlead) [inline src/Cake/Program.cs]: this needed? :wink:

File: src/Cake/Scripting/XPlat/XPlatScriptSession.cs [hunk 2]
```diff
@@ -46,8 +46,14 @@ public void AddReference(FilePath path)
             {
                 throw new ArgumentNullException(nameof(path));
             }
+
+#if NETCORE
+            var assembly = Assembly.Load(new AssemblyName(path.FullPath));
+            References.Add(assembly);
+#else
             _log.Debug("Adding reference to {0}...", path.GetFilename().FullPath);
```
Comment (devlead) [inline src/Cake/Scripting/XPlat/XPlatScriptSession.cs]: Could we move the log above the if def? Good info to have.

File: src/Cake/Scripting/XPlat/XPlatScriptSession.cs [hunk 3]
```diff
@@ -46,8 +46,14 @@ public void AddReference(FilePath path)
             {
                 throw new ArgumentNullException(nameof(path));
             }
+
             _log.Debug("Adding reference to {0}...", path.GetFilename().FullPath);
+#if NETCORE
+            var assembly = Assembly.Load(new AssemblyName(path.FullPath));
```
Comment (patriksvensson) [inline src/Cake/Scripting/XPlat/XPlatScriptSession.cs]: Use assembly loader here.

File: src/Cake.Core/Scripting/ScriptConventions.cs [hunk 4]
```diff
@@ -49,33 +66,57 @@ public IReadOnlyList<string> GetDefaultNamespaces()
...
+            foreach (var cakeAssembly in cakeAssemblies)
             {
-                typeof(Action).GetTypeInfo().Assembly, // mscorlib
-                typeof(Uri).GetTypeInfo().Assembly, // System
-                typeof(IQueryable).GetTypeInfo().Assembly, // System.Core
-#if !NETCORE
-                typeof(System.Data.DataTable).GetTypeInfo().Assembly, // System.Data
+                foreach (var reference in cakeAssembly.GetReferencedAssemblies())
+                {
+                    result.Add(Assembly.Load(reference));
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/ScriptConventions.cs]: Use assembly loader.

**Comment by devlead** [comment]
LGTM :+1: Tested both `#addin` and `#r` without issue, I'll do some more testing and then I'll merge.

**Pull request merged**